### PR TITLE
PR for C++ training exercises

### PR DESCRIPTION
**Pull request by BhuvanBezawada**
**Report to:** @LamarMoore 

**PR Details:**
Finished both the training exercises and submitting the PR as part of training.

### CODE REVIEW

**Comment by LamarMoore** [comment]
Doesn't build on Windows. Can you fix this error?

File: exercises/bhuvan_bezawada/ex01_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex01_basics/src/main.cpp]: There isn't any error handling around this. Would be nice if the user is notified that they have entered an invalid filename or event an empty file.

File: exercises/bhuvan_bezawada/ex01_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex01_basics/src/main.cpp]: I guess strictly speaking you would want to parse command-line arguments. For example what is shown here [Parsing Command-Line Arguments](http://www.cplusplus.com/articles/DEN36Up4/)

File: exercises/bhuvan_bezawada/ex01_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex01_basics/src/main.cpp]: Why do you need to check for spaces here? Aren't they all removed by `split()`?

File: exercises/bhuvan_bezawada/ex01_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex01_basics/src/main.cpp]: In general it may not be a good idea to change the size of an array whilst looping over it. Also, hyphenated words are treated as separate words, so do you need to remove the hyphen? What does `ispunct()` return for a hyphen? Also, shouldn't you create a two new words from words separated by punctuation?

File: exercises/bhuvan_bezawada/ex01_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex01_basics/src/main.cpp]: 👍

File: exercises/bhuvan_bezawada/ex01_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex01_basics/src/main.cpp]: You might be able to avoid this loop since it is most likely punctuation will appear at the end of the word.

File: exercises/bhuvan_bezawada/ex01_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex01_basics/src/main.cpp]: A message to the user that you have written an output file and its location would be friendlier.

**Review by LamarMoore — CHANGES_REQUESTED** [review]

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp]: Normally, we like to separate the implementation from the definition. e.g `Shape.h, Shape.cpp` and then we include the  header files we need to use. You do not need to change this now but bear this in mind moving forward.

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp]: Do these need to be public and modifiable to the outside world? Consider making these public and readable via getter methods.

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp]: Can `num_sides` be < 0? Maybe `unsigned int` or even `unsigned __int16`

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp]: classes with pure virtual methods which are intended to act as base classes should have a virtual destructor.

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp]: This could be part of the `Shape` constructor. E.g `Shape(const std::string &type, int numsides)` which then sets the shape and you could do something like `Square(int side) : Shape("square", 4)`. Have you considered using an `enum class` for the type or some other type instead of a `std::string` for storing the type?

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp]: add the `override` keyword to overidden methods e.e `double perimeter() override {}`

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp]: Why capture as `int` and store as `double`? Why not capture as `double` in the first place. Also `side1` is a bad name. What about `width` or `sideLength`? You may think of something better.

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp]: `print_type()` and `print_num_sides()` do not need to be redefined if they exist in the base class. These methods are essentially duplicated. Also, are these required based on the requirements in the question?

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp]: Instead of doing this, you could have created a vector of shapes and just taken a reference to the vector in all of your `ShapeSorter` methods. Instead you are copying the pointer vector everytime you call a method, then copying again after you enter.

**Review by LamarMoore — CHANGES_REQUESTED** [review]
Hi a number of things here I highlighted which could improve your inheritance structure and design in general.

File: exercises/bhuvan_bezawada/ex01_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex01_basics/src/main.cpp]: indentation (and the two if blocks below this)

File: exercises/bhuvan_bezawada/ex01_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex01_basics/src/main.cpp]: you should return/exit here too. Maybe just return non zero

File: exercises/bhuvan_bezawada/ex01_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex01_basics/src/main.cpp]: return non zero?

File: exercises/bhuvan_bezawada/ex01_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex01_basics/src/main.cpp]: indentation

File: exercises/bhuvan_bezawada/ex01_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex01_basics/src/main.cpp]: Still not clear why you are deleting the hyphen. Are you deleting the hyphen?

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp]: indentation. Consider download the LLVM clang-format plugin for visual studio.

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp]: indentation. Review all your code and make sure your indentation is correct.

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp]: width, height as meaningful names ?

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp]: Still not sure why these print methods are necessary.

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (LamarMoore) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp]: spell out what the method is doing in the name.

**Review by LamarMoore — COMMENTED** [review]
Mostly minor comments surrounding indentation etc. When naming methods try to be explicit so that someone reading the interface will understand what the method is actually doing. Also, the naming styles which should be used can be found here [Coding Standards](http://developer.mantidproject.org/Standards/index.html). All developers abide by these standards so that we can have consistency in the codebase and we all know what to expect when reading each others code.

**Comment by LamarMoore** [comment]
Once the RHEL build is working I will approve these changes.

**Review by LamarMoore — APPROVED** [review]

File: exercises/bhuvan_bezawada/ex01_basics/src/main.cpp
Comment (OwenArnold) [inline exercises/bhuvan_bezawada/ex01_basics/src/main.cpp:40]: This is acceptable for the purpose of the exercises, but you should be aware of `regular expressions` if you are not already as finding words would be much easier using them.

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (OwenArnold) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp:42]: I would not hard-code the IO like this, I would return a string and leave it up to the client code to determine how to best present/persist it.

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (OwenArnold) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp:47]: Same as above

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (OwenArnold) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp:62]: Isn't a square a special case of a rectangle? Did you consider having Square inherit from Rectangle instead?

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (OwenArnold) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp:38]: `perimeter` and `area` should be `const`. They should guarantee not making changes to internal state.

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (OwenArnold) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp:187]: As a word of warning, careful with `std::endl` in for loops. It involves an expensive flush.

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (OwenArnold) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp:193]: Why pass `shapes` by value instead of by const ref?

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (OwenArnold) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp:195]: passing the function predicate as a named function is fine, you could also have done this with a lambda, though I prefer your approach for readability.

File: exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp
Comment (OwenArnold) [inline exercises/bhuvan_bezawada/ex02_oo_basics/src/main.cpp:231]: Interesting, you have no memory leak here so technically what you have done is fine, but the vector `shapes` does not own the shapes, and that isn't terribly clear. Maybe these should be stored as const pointers in the vector? Otherwise you could have stored them as `std::unique_ptr`

**Review by OwenArnold — COMMENTED** [review]
Looks pretty good to me. I've made some comments for you to think about.

**Pull request closed without merging**
